src/metabase/sync/analyze/fingerprint/fingerprinters.clj

```diff
@@ -224,11 +223,27 @@
         :q1  q1
         :q3  q3)))))
 
+(def truncation-size
+  "The maximum size of :type/Text to be selected from the database in `table-rows-sample`. In practice we see large
+  text blobs and want to balance taking enough for distinct counts and but not so much that we risk out of memory
+  issues when syncing."
+  1234)
+
 (defn- valid-serialized-json?
-  "Is x a serialized JSON dictionary or array."
+  "Is x a serialized JSON dictionary or array. Hueristically recognize maps and arrays. Uses the following strategies:
+  - leading character {: assume valid JSON
+  - leading character [: assume valid json unless its of the form [ident] where ident is not a boolean."
   [x]
   (u/ignore-exceptions
-    ((some-fn map? sequential?) (json/parse-string x))))
+    (when (and x (string? x))
+      (let [matcher (case (first x)
```

Comment:
Is this really measurably more performant than a direct/naive regex implementation? I'd just do the first 2 characters without too many assumptions. Look at it like this: it will be faster than what we have, so it's a win. From here I'd optimize for readability & robustness.

Reply:
i was going to do the first two characters but then we would miss out on `"[true, true]"` is valid JSON but `"[prod]"` is not valid json. that's the only reason. if we want to ignore that case its fine. As I understand it we only care about the first character if its `{` and treat that as a valid map. If its a vector you had mentioned checking its not a single `[tag]` style text.

Reply:
Fair point. Keep in mind though that we require 95% of rows to be valid, so false positives are almost always fine, and we can tolerate at least some false negatives.